test/unit/org/apache/cassandra/index/sai/utils/PrimaryKeyTest.java

```diff
@@ -360,22 +348,4 @@ private void compareToAndEqualsTests(PrimaryKey.Factory factory, PrimaryKey... k
             }
         }
     }
-
-    private void byteComparisonTests(PrimaryKey.Factory factory, PrimaryKey... keys)
-    {
-        for (int index = 0; index < keys.length - 1; index++)
-        {
-            PrimaryKey key = keys[index];
-            PrimaryKey tokenOnlyKey = factory.createTokenOnly(key.token());
-            assertByteComparison(tokenOnlyKey, key, -1);
-            assertByteComparison(key, key, 0);
-            assertByteComparison(tokenOnlyKey, tokenOnlyKey, 0);
-
-            for (int comparisonIndex = index + 1; comparisonIndex < keys.length; comparisonIndex++)
-            {
-                assertByteComparison(key, keys[comparisonIndex], -1);
-                assertByteComparison(tokenOnlyKey, keys[comparisonIndex], -1);
-            }
-        }
-    }
```

Comment:
So I'm trying to understand this in the context of the changes in `PrimaryKey`, and it might just be late, but I'm a bit confused. Instead of deleting this, I tried something else...

```
private void byteComparisonTests(PrimaryKey.Factory factory, PrimaryKey... keys)
{
    for (int index = 0; index < keys.length - 1; index++)
    {
        PrimaryKey key = keys[index];
        assertByteComparison(key, key, 0);

        for (int comparisonIndex = index + 1; comparisonIndex < keys.length; comparisonIndex++)
        {
            assertTrue(compareTerms(readBytes(key), readBytes(keys[comparisonIndex])) < 0);
        }
    }
}

private int compareTerms(BytesRef left, BytesRef right)
{
    return FastByteOperations.compareUnsigned(left.bytes, left.offset, left.offset + left.length,
                                              right.bytes, right.offset, right.offset + right.length);
}

private BytesRef readBytes(ByteComparable source)
{
    BytesRefBuilder builder = new BytesRefBuilder();

    ByteSource byteSource = source.asComparableBytes(ByteComparable.Version.OSS50);
    int val;
    while ((val = byteSource.next()) != ByteSource.END_OF_STREAM)
        builder.append((byte) val);
    return builder.get();
}
```
This seems to be more or less what `SortedTermsReader` does, and we've got a sequence of strictly increasing integers for our simple primary keys, but it fails. Is there no way to compare them in unsigned byte space?

Reply:
Okay, so I think I see what happened! The sort takes the token into account, and that reorders the byte-comparables. So you could do this, and it would work just fine:

```
@Test
    public void singlePartitionTest()
    {
        PrimaryKey.Factory factory = new PrimaryKey.Factory(simplePartition.comparator);
        int rows = nextInt(10, 100);
        PrimaryKey[] keys = new PrimaryKey[rows];
        for (int index = 0; index < rows; index++)
            keys[index] = factory.create(makeKey(simplePartition, index), Clustering.EMPTY);

        byteComparisonTests(keys);
        Arrays.sort(keys);
        compareToAndEqualsTests(factory, keys);
    }
```
...and then...
```
private void byteComparisonTests(PrimaryKey... keys)
    {
        for (int index = 0; index < keys.length - 1; index++)
        {
            PrimaryKey key = keys[index];
            assertByteComparison(key, key, 0);

            for (int comparisonIndex = index + 1; comparisonIndex < keys.length; comparisonIndex++)
            {
                assertByteComparison(key, keys[comparisonIndex], -1);
            }
        }
    }
```
I suppose your removal says that you felt this wasn't really that useful, given we're not in the business of testing byte-comparable types themselves, and our usage of this is covered in `SortedTermsTest`?

Reply:
I tried various options here but, as you say, we aren't really in the business of testing byte comparables so it makes no sense to try and do it here. By removing the token from the byte comparable, the primary keys no longer maintain any order. 

I have considered not using byte comparable at all for the primary keys and this may be option for the future. We could just serialize the keys and use byte buffers in place of the comparables. This would allow us to remove some of the array copying that we currently do and remove our dependency of the `BytesRef` in the sorted terms.

Reply:
> I have considered not using byte comparable at all for the primary keys and this may be option for the future.

I'd wondered that reading this patch as well. Assuming it doesn't have any real impact on compression ratio, that would be the way to go. Not opposed to it being a separate issue, but you can also make a case we never would have used byte-comparable stuff without the trie. To someone not familiar w/ the evolution, it will look a bit odd :D

Reply:
There is this bit in the package info...

```
 * Terms are stored in <code>ByteComparable</code> strictly ascending order.
 * Duplicates are not allowed.
```

Using `ByteComparable` does ensure that the values are ordered by their unsigned representation, and `SortedTermsReader` relies on that, right?

Reply:
That bit needs to come out. Neither the partition keys nor the clustering keys are in any order, and we do now allow duplicates.

Reply:
I would prefer not to remove the `ByteComparable` in this patch. I'd like to combine that in another patch along with removing our reliance on Lucene `DataOutput` and `IndexOutput`. I will raise a ticket to remove this usage so that we can write `ByteBuffer`s directly without having to move them around.

Reply:
> I will raise a ticket to remove this usage so that we can write ByteBuffers directly without having to move them around.

+1